mcp/streamable.go

```diff
@@ -390,47 +443,93 @@ func (t *StreamableServerTransport) servePOST(w http.ResponseWriter, req *http.R
 			// the HTTP request. If we didn't do this, a request with a bad method or
 			// missing ID could be silently swallowed.
 			if _, err := checkRequest(req, serverMethodInfos); err != nil {
-				return http.StatusBadRequest, err.Error()
+				http.Error(w, err.Error(), http.StatusBadRequest)
+				return
 			}
 			if req.ID.IsValid() {
 				requests[req.ID] = struct{}{}
 			}
 		}
 	}
 
-	// Update accounting for this request.
-	stream := newStream(StreamID(t.nextStreamID.Add(1)))
-	t.mu.Lock()
-	t.streams[stream.id] = stream
+	var stream *stream // if non-nil, used to handle requests
+
+	// If we have requests, we need to handle responses along with any
+	// notifications or server->client requests made in the course of handling.
+	// Update accounting for this incoming payload.
 	if len(requests) > 0 {
-		stream.requests = make(map[jsonrpc.ID]struct{})
-	}
-	for reqID := range requests {
-		t.requestStreams[reqID] = stream.id
-		stream.requests[reqID] = struct{}{}
+		stream = newStream(StreamID(t.lastStreamID.Add(1)), t.opts.JSONResponse)
+		t.mu.Lock()
+		t.streams[stream.id] = stream
+		stream.requests = requests
+		for reqID := range requests {
+			t.requestStreams[reqID] = stream.id
+		}
+		t.mu.Unlock()
+		stream.signal.Store(signalChanPtr())
 	}
-	t.mu.Unlock()
-	stream.signal.Store(signalChanPtr())
 
 	// Publish incoming messages.
 	for _, msg := range incoming {
 		t.incoming <- msg
 	}
 
-	// TODO(rfindley): consider optimizing for a single incoming request, by
-	// responding with application/json when there is only a single message in
-	// the response.
-	// (But how would we know there is only a single message? For example, couldn't
-	//  a progress notification be sent before a response on the same context?)
-	return t.streamResponse(stream, w, req, -1)
+	if stream == nil {
+		w.WriteHeader(http.StatusAccepted)
+		return
+	}
+
+	if stream.jsonResponse {
```

Comment:
Always use respondJSON here for the optimization

Reply:
FWIW, that's now quite right: we actually want to do this in handleSSE, if and only if we only get one response message. We don't know that yet, here.

I'll do this as a separate PR.